components/p3a/brave_p3a_star_log_store.cc

```diff
@@ -0,0 +1,194 @@
+  std::string epoch_key = base::NumberToString(epoch);
+  if (update->FindDictKey(epoch_key) == nullptr) {
+    update->SetKey(epoch_key, base::Value(base::Value::Type::DICT));
+  }
+  update->FindDictKey(epoch_key)->SetStringKey(histogram_name, msg);
+
+  if (current_epoch_ != epoch) {
+    LogKey key(epoch, histogram_name);
+    log_[key] = msg;
+    unsent_entries_.insert(key);
+  }
+}
+
+void BraveP3AStarLogStore::RemoveMessageIfExists(const LogKey& key) {
+  log_.erase(key);
+  unsent_entries_.erase(key);
+
+  // Update the persistent value.
+  DictionaryPrefUpdate update(local_state_, kPrefName);
+  std::string epoch_key = base::NumberToString(key.epoch);
+  base::Value* epoch_dict = update->FindDictKey(epoch_key);
+  if (epoch_dict != nullptr) {
+    epoch_dict->RemoveKey(key.histogram_name);
+  }
+
+  if (has_staged_log() && staged_entry_key_->epoch == key.epoch &&
+      staged_entry_key_->histogram_name == key.histogram_name) {
+    staged_entry_key_.reset();
+    staged_log_.clear();
+  }
+}
+
+void BraveP3AStarLogStore::SetCurrentEpoch(uint8_t current_epoch) {
+  current_epoch_ = current_epoch;
+}
+
+bool BraveP3AStarLogStore::has_unsent_logs() const {
+  return !unsent_entries_.empty();
+}
+
+bool BraveP3AStarLogStore::has_staged_log() const {
+  return staged_entry_key_ != nullptr;
+}
+
+const std::string& BraveP3AStarLogStore::staged_log() const {
+  DCHECK(staged_entry_key_);
+  DCHECK(!staged_log_.empty());
+
+  return staged_log_;
+}
+
+std::string BraveP3AStarLogStore::staged_log_type() const {
+  DCHECK(staged_entry_key_);
+  if (base::StartsWith(staged_entry_key_->histogram_name, "Brave.P2A",
+                       base::CompareCase::SENSITIVE)) {
+    return "p2a";
+  }
+  return "p3a";
+}
+
+const std::string& BraveP3AStarLogStore::staged_log_hash() const {
+  NOTREACHED();
+  return staged_log_hash_;
+}
+
+const std::string& BraveP3AStarLogStore::staged_log_signature() const {
+  NOTREACHED();
+  return staged_log_signature_;
+}
+
+absl::optional<uint64_t> BraveP3AStarLogStore::staged_log_user_id() const {
+  NOTREACHED();
+  return absl::nullopt;
+}
+
+void BraveP3AStarLogStore::StageNextLog() {
+  // Stage the next item.
+  DCHECK(has_unsent_logs());
+  uint64_t rand_idx = base::RandGenerator(unsent_entries_.size());
+  staged_entry_key_.reset(new LogKey(*(unsent_entries_.begin() + rand_idx)));
+
+  staged_log_ = log_.at(*staged_entry_key_);
+
+  VLOG(2) << "BraveP3AStarLogStore::StageNextLog: staged epoch = "
+          << static_cast<int>(staged_entry_key_->epoch)
+          << ", histogram_name = " << staged_entry_key_->histogram_name;
+}
+
+void BraveP3AStarLogStore::DiscardStagedLog() {
+  if (!has_staged_log()) {
+    return;
+  }
+  staged_entry_key_.reset();
+  staged_log_.clear();
+}
+
+void BraveP3AStarLogStore::MarkStagedLogAsSent() {
+  if (!has_staged_log()) {
+    return;
+  }
+  RemoveMessageIfExists(*staged_entry_key_);
+}
+
+void BraveP3AStarLogStore::TrimAndPersistUnsentLogs() {
+  NOTREACHED();
```

Comment:
Judging from Chromium comments, `NOTIMPLEMENTED` is for codepaths that have not been implemented **yet**. Since there are no plans to implement these methods in the future, `NOTREACHED` might be a better choice.